java/java-analysis-api/src/com/intellij/lang/jvm/actions/AnnotationAttributeValueRequest.kt

```diff
@@ -1,24 +1,81 @@
-  ```
-  they are skipped until become necessary
-   */
-
+  data class ClassValue(val classFqn: String) : AnnotationAttributeValueRequest()
+  data class ConstantValue(val reference: JvmField?, val text: String) : AnnotationAttributeValueRequest()
+  data class NestedAnnotation(val annotationRequest: AnnotationRequest) : AnnotationAttributeValueRequest()
+  data class ArrayValue(val members: List<AnnotationAttributeValueRequest>) : AnnotationAttributeValueRequest()
 }
 
 data class AnnotationAttributeRequest(val name: String, val value: AnnotationAttributeValueRequest)
 
-fun stringAttribute(name: String, value: String): AnnotationAttributeRequest = AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.StringValue(value))
-fun intAttribute(name: String, value: Int): AnnotationAttributeRequest = AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.PrimitiveValue(value))
+fun stringAttribute(name: String, value: String): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.StringValue(value))
+
+fun intAttribute(name: String, value: Int): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.PrimitiveValue(value))
+
+fun classAttribute(name: String, classFqn: String): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.ClassValue(classFqn))
+
+fun constantAttribute(name: String, reference: JvmField): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.ConstantValue(reference, (reference as PsiElement).text))
+
+fun constantAttribute(name: String, refText: String): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.ConstantValue(null, refText))
+
+fun nestedAttribute(name: String, annotation: AnnotationRequest): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.NestedAnnotation(annotation))
+
+fun arrayAttribute(name: String, members: List<AnnotationAttributeValueRequest>): AnnotationAttributeRequest =
+  AnnotationAttributeRequest(name, AnnotationAttributeValueRequest.ArrayValue(members))
+
+fun JvmAnnotationAttributeValue.toAttributeValueRequest(attribute: JvmAnnotationAttribute? = null): AnnotationAttributeValueRequest? = when (this) {
+  is JvmAnnotationArrayValue -> AnnotationAttributeValueRequest.ArrayValue(
+    this.values.mapNotNull { it.toAttributeValueRequest(attribute) })
+  is JvmAnnotationClassValue -> {
+    val referenceText = this.clazz?.qualifiedName ?: run {
+      // Try to preserve unresolved class literal
+      val classLiteralExpression = ((attribute as? PsiNameValuePair)?.value as? PsiExpression)?.type
+      val genericParameter = (classLiteralExpression as? PsiClassType)?.parameters?.firstOrNull()
+      genericParameter?.canonicalText
+    }
```

Comment:
Well even if it will probably not be useful most of the time, the fact that unresolved symbols are removed when invoking the action is bothering me a bit.

It's the biggest "subject of conflict" I had when writing these changes. If you do not consider it important enough I can remove it and simply discard unresolved symbols. My original problem was that annotations couldn't be kept when invoking the action after.


Also I don't see how I could extract the "fallback handling" code to another function.


The thing I (kinda) forgot about the Kotlin plugin is that it needs to be updated too (I get an exception right now because of the unhandled attributes.)
I cloned the `intellij-kotlin` repo and got it running so I'll do my best to send a PR there when I get the time to (I hope for this weekend.)

Reply:
> If you do not consider it important enough I can remove it and simply discard unresolved symbols

Actually it is really hard to make refactoring work properly on unresolved code, and it is nearly impossible when we are talking about inter-language refactorings. So I think that here it just doesn't worth it. So if you haven't any specific case for it then I think it will be better to drop it.

> Also I don't see how I could extract the "fallback handling" code to another function.

I meant something like:

```
private fun tryHandleUnresolvedFQN(attribute: JvmAnnotationAttribute?): @NlsSafe String? {
    // Try to preserve unresolved class literal
    val classLiteralExpression = ((attribute as? PsiNameValuePair)?.value as? PsiExpression)?.type
    val genericParameter = (classLiteralExpression as? PsiClassType)?.parameters?.firstOrNull()
   return genericParameter?.canonicalText
}
```